core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -523,15 +523,26 @@ static void checkForIndexDataInDefaultPathData(
 
         boolean clean = true;
         for (final String defaultPathData : Environment.DEFAULT_PATH_DATA_SETTING.get(settings)) {
-            final Path nodeDirectory = NodeEnvironment.resolveNodePath(getPath(defaultPathData), nodeEnv.getNodeLockId());
-            if (Files.exists(nodeDirectory) == false) {
+            final Path defaultNodeDirectory = NodeEnvironment.resolveNodePath(getPath(defaultPathData), nodeEnv.getNodeLockId());
+            if (Files.exists(defaultNodeDirectory) == false) {
                 continue;
             }
-            final NodeEnvironment.NodePath nodePath = new NodeEnvironment.NodePath(nodeDirectory);
+
+            boolean includedInPathData = false;
+            for (final NodeEnvironment.NodePath dataPath : nodeEnv.nodePaths()) {
+                includedInPathData |= Files.isSameFile(dataPath.path, defaultNodeDirectory);
```

Comment:
maybe move the loop in a method then we can return early and don't need a label / boolean var in the outer loop